libsolidity/ast/AST.h

```diff
@@ -721,6 +717,35 @@ class VariableDeclaration: public Declaration
 	Location m_location; ///< Location of the variable if it is of reference type.
 };
 
+class ModifierArea: public ASTNode
+{
+public:
+	ModifierArea(
+		SourceLocation const& _location,
+		std::vector<ASTPointer<ModifierInvocation>> const& _modifiers,
+		ASTPointer<std::vector<ASTPointer<FunctionDefinition>>> const& _functions,
+		ASTPointer<std::vector<ASTPointer<ModifierArea>>> const& _subAreas,
+		ASTPointer<ModifierArea> const& _parent = ASTPointer<ModifierArea>()
+	);
+
+	virtual void accept(ASTVisitor& _visitor) override;
+	virtual void accept(ASTConstVisitor& _visitor) const override;
+
+	/// Returns only modifiers defined
+	std::vector<ASTPointer<ModifierInvocation>> const& definedModifiers() const { return m_declaredModifiers; }
+	/// Returns all modifiers in effect (including ones from parents)
+	std::vector<ASTPointer<ModifierInvocation>> const& totalModifiers() const { return m_totalModifiers; }
+	std::vector<ASTPointer<FunctionDefinition>> const& definedFunctions() const { return *m_functions; }
+	std::vector<ASTPointer<ModifierArea>> const& subAreas() const { return *m_subAreas; }
+
+private:
+	std::vector<ASTPointer<ModifierInvocation>> m_totalModifiers;
+	std::vector<ASTPointer<ModifierInvocation>> m_declaredModifiers;
+	ASTPointer<std::vector<ASTPointer<FunctionDefinition>> const> m_functions;
```

Comment:
Actually: it doesn't seem like those can ever be null or that they are passed around, resp. shared - so why use ``ASTPointer``s and not just the ``vector``s themselves as members?